src/vivarium/framework/lifecycle/manager.py

```diff
@@ -0,0 +1,280 @@
+
+        Returns
+        -------
+            A list of state names in order of execution.
+        """
+        return self.lifecycle.get_state_names(phase)
+
+    def add_handlers(self, state_name: str, handlers: list[Callable[[Event], None]]) -> None:
+        """Registers a set of functions to be called during a life cycle state.
+
+        This method does not apply any constraints, rather it is used
+        to build up an execution order for introspection.
+
+        Parameters
+        ----------
+        state_name
+            The name of the state to register the handlers for.
+        handlers
+            A list of functions that will execute during the state.
+        """
+        s = self.lifecycle.get_state(state_name)
+        s.add_handlers(handlers)
+
+    def add_constraint(
+        self,
+        method: Callable[..., Any],
+        allow_during: tuple[str, ...] | list[str] = (),
+        restrict_during: tuple[str, ...] | list[str] = (),
+    ) -> None:
+        """Constrains a function to be executable only during certain states.
+
+        Parameters
+        ----------
+        method
+            The method to add constraints to.
+        allow_during
+            An optional list of life cycle states in which the provided
+            method is allowed to be called.
+        restrict_during
+            An optional list of life cycle states in which the provided
+            method is restricted from being called.
+
+        Raises
+        ------
+        ValueError
+            If neither ``allow_during`` nor ``restrict_during`` are provided,
+            or if both are provided.
+        LifeCycleError
+            If states provided as arguments are not in the life cycle.
+        ConstraintError
+            If a lifecycle constraint has already been applied to the provided
+            method.
+        """
+        if allow_during and restrict_during or not (allow_during or restrict_during):
+            raise ValueError(
+                'Must provide exactly one of "allow_during" or "restrict_during".'
+            )
+        unknown_states = (
+            set(allow_during).union(restrict_during).difference(self.lifecycle._state_names)
+        )
+        if unknown_states:
+            raise LifeCycleError(
+                f"Attempting to constrain {method} with "
+                f"states not in the life cycle: {list(unknown_states)}."
+            )
+        if restrict_during:
+            allow_during = [
+                s for s in self.lifecycle._state_names if s not in restrict_during
+            ]
+        if isinstance(allow_during, tuple):
+            allow_during = list(allow_during)
+
+        self._make_constraint(method, allow_during)
+
+    def __repr__(self) -> str:
+        return f"LifeCycleManager(state={self.current_state})"
+
+    def __str__(self) -> str:
+        return str(self.lifecycle)
+
+
+class LifeCycleInterface(Interface):
```

Comment:
We're inconsistent in whether we break interface and manager into separate modules. What do we prefer?

Reply:
I support being consistent.